drivers/clock_control/nrf_power_clock.c

```diff
@@ -9,272 +9,296 @@
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
+#include <logging/log.h>
 #include <nrf_power.h>
-#include <drivers/clock_control/nrf_clock_control.h>
-#endif
 
-static u8_t m16src_ref;
-static u8_t m16src_grd;
-static u8_t k32src_initialized;
+LOG_MODULE_REGISTER(clock_control, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+/* Helper logging macros which prepends device name to the log. */
+#define CLOCK_LOG(lvl, dev, ...) \
+	LOG_##lvl("%s: " GET_ARG1(__VA_ARGS__), dev->config->name \
+			COND_CODE_0(NUM_VA_ARGS_LESS_1(__VA_ARGS__),\
+					(), (, GET_ARGS_LESS_1(__VA_ARGS__))))
+#define ERR(dev, ...) CLOCK_LOG(ERR, dev, __VA_ARGS__)
+#define WRN(dev, ...) CLOCK_LOG(WRN, dev, __VA_ARGS__)
+#define INF(dev, ...) CLOCK_LOG(INF, dev, __VA_ARGS__)
+#define DBG(dev, ...) CLOCK_LOG(DBG, dev, __VA_ARGS__)
+
+/* returns true if clock stopping must be deferred due to active calibration. */
+typedef void (*nrf_clock_stop_handler_t)(struct device *dev);
+
+/* Clock instance structure */
+struct nrf_clock_control {
+	s32_t ref;		/* Users counter */
+	sys_slist_t list;	/* List of users requesting callback */
+	bool started;		/* When set clock is started */
+};
 
-static int m16src_start(struct device *dev, clock_control_subsys_t sub_system)
-{
-	bool blocking;
-	u32_t imask;
-	u32_t stat;
+/* Clock instance static configuration */
+struct nrf_clock_control_config {
+	u32_t start_tsk;	/* Clock start task */
+	u32_t started_evt;	/* Clock started event */
+	u32_t stop_tsk;		/* Clock stop task */
+	nrf_clock_stop_handler_t stop_handler; /* Function called before stop */
+};
 
-	/* If the clock is already started then just increment refcount.
-	 * If the start and stop don't happen in pairs, a rollover will
-	 * be caught and in that case system should assert.
-	 */
+#define INT_MASK \
+	(NRF_CLOCK_INT_HF_STARTED_MASK | NRF_CLOCK_INT_LF_STARTED_MASK | \
+	COND_CODE_1(CONFIG_USB_NRF52840, \
+		    (NRF_POWER_INT_USBDETECTED_MASK | \
+		     NRF_POWER_INT_USBREMOVED_MASK | \
+		     NRF_POWER_INT_USBPWRRDY_MASK), (0)))
 
-	/* Test for reference increment from zero and resource guard not taken.
-	 */
-	imask = irq_lock();
+/* Return true if given event has enabled interrupt and is triggered. Event
+ * is cleared.
+ */
+static bool clock_event_check_and_clean(u32_t evt, u32_t intmask)
+{
+	bool ret = nrf_clock_event_check(evt) &&
+			nrf_clock_int_enable_check(intmask);
 
-	if (m16src_ref++) {
-		irq_unlock(imask);
-		goto hf_already_started;
+	if (ret) {
+		nrf_clock_event_clear(evt);
 	}
 
-	if (m16src_grd) {
-		m16src_ref--;
-		irq_unlock(imask);
-		return -EAGAIN;
-	}
+	return ret;
+}
+
+static enum clock_control_status get_status(struct device *dev,
+					    clock_control_subsys_t sys)
+{
+	struct nrf_clock_control *data = dev->driver_data;
 
-	m16src_grd = 1U;
+#if 0
+	__ASSERT_NO_MSG(z_arm_irq_current_prio() >
+			NVIC_GetPriority(nrfx_get_irq_number(NRF_CLOCK)));
+#endif
 
-	irq_unlock(imask);
+	if (data->ref > 0) {
+		return data->started ?
```

Comment:
That's true but on the other hand one can question that result of get status is always volatile if someone that did not requested clock is checking it's state. Main purpose for having that function is to block until clock is ready. In that case if one requested clock then it will never be in `OFF` state.

I wouldn't like to complicate this function only to satisfy use case which will rather never occur. Maybe we should clarify `clock_control.h` the use case. Maybe we should even replace this function with something like `clock_control_wfo()` (wait for on)?

Reply:
What about changing it to:
```
if (data->started) {
	return CLOCK_CONTROL_STATUS_ON;
}
if (data->ref > 0) {
	return CLOCK_CONTROL_STATUS_STARTING;
}

return CLOCK_CONTROL_STATUS_OFF;
```